### ISSUE TITLE
asInt and asFloat

### ISSUE THREAD

**Issue by jumpjack** [adventuregamestudio/ags#1868]
I am getting weird results using .asFloat :

- "1234567890" becomes 1234567936.000
- "123456789.123" becomes 123456792.00
- "123453.1234" becomes 123456.125

Also weird results with asInt:

- "2147483701" becomes 2147483647
- "2147483674" becomes 2147483647 too.

Am I using them wrong, or is this a bug?

Using AGS 3.6.0 RC3.

What I need to do is subtracting 2^32 - 1 from a number, but i'm getting overflow so I switched to float; I can't figure out how many bits AGS is using for integers representation.

My code:

 ```
 s = "2147483701";
  f=s.AsFloat;
  L=s.AsInt;
  i=s.AsInt;
debugPrint(String.Format("string %s to float '%f' and long %d and int %d", s, f,  L , i),true);  

  s = "2147483674";
  f=s.AsFloat;
  L=s.AsInt;
debugPrint(String.Format("string %s to float '%f' and long %d and int %d", s, f,  L , i),true);  

  s = "123456.1234";
  f=s.AsFloat;
  L=s.AsInt;
debugPrint(String.Format("string %s to float '%f' and long %d and int %d", s, f,  L , i),true)

```

**Comment by ivan-mogilko** [comment]
AGS script works only with signed 32-bit numbers, so integers may have range of -2147483648 to 2147483647. So with integers you definitely have an overflow.

I'm not certain about floats yet, I'll need to investigate.

**Comment by jumpjack** [comment]
> AGS script works only with signed 32-bit numbers, so integers may have range of -2147483648 to 2147483647. So with integers you definitely have an overflow in above examples.
> 

Is it possible to add this information somewhere in the manual? By sure in asInt and asFloat, but maybe also [StringFormats](https://adventuregamestudio.github.io/ags-manual/StringFormats.html) and  [StandardTypes ](https://adventuregamestudio.github.io/ags-manual/StandardTypes.html) would be good places where to put the info.

(p.s.: yes I ma de a typo).

**Comment by ericoporto** [comment]
It's here: https://adventuregamestudio.github.io/ags-manual/ScriptKeywords.html#data-types

**Comment by jumpjack** [comment]
I also found a "long", but...

![image](https://user-images.githubusercontent.com/1620953/209331313-57f24cfb-c15a-4adc-b492-2017d91c0cc3.png)

 ```
  long big = 2^32;
  int small = 2^32;
  debugPrint(String.Format("%ld, %lld, %d", big,big,   small), true);
  
  big = 1;
  small = 1;
  debugPrint(String.Format("%ld, %lld, %d", big,big,   small), true);

  big = -1;
  small = -1;
  debugPrint(String.Format("%ld, %lld, %d", big,big,   small), true);  
```

results in:

- 34, 5898828343410722, 34 
- 1, 5898828343410689, 1   
- -1, 5898832638377983, -1

**Comment by ericoporto** [comment]
Long is probably there by mistake of the update in the Scintilla component of the Editor - the text editor used.

**Comment by jumpjack** [comment]
I am very confused now:

  ```
float big = Maths.RaiseToPower(2.0 , 32.0); // 4294967296.0000
  float t = 2147483676.0000; // 2147483648.000000 ??? (=2^32/2)
  float result = big - t;
  debugPrint(String.Format("%f - %f = %f ?", big,t, result), true);
  
  long big2 = 2^32; // 4294967296.0000
  debugPrint(String.Format("%ld, %lld, %f, %Lf", big2, big2, big2, big2), true); 

```

**Result:**
- 4294967296.000000 - 2147483648.000000 = 2147483648.000000 ? 
- 34, 5898742444064802, 0.000000, 0.000000

**Comment by ivan-mogilko** [comment]
I must underline again: AGS script does not support 64-bit numbers, only signed 32-bit, so using `long` won't change anything, it's a equivalent to `int`. I was not even aware of its existance tbh.

Because the numbers are 32-bit, using `%lld` formatting possibly breaks the printf function.

**Comment by jumpjack** [comment]
Yes, but also look at first result:

float big = Maths.RaiseToPower(2.0 , 32.0)  --> ok, it gives 4294967296.0000

float t = 2147483676.0000  --->**KO , turns into 2^32/2**  ( 2147483648.000000  )


I cannot locate the bug.

Other tests:

  ```
float t;
  t = 2148000000.0000;   
  debugPrint(String.Format("2148000000 %f", t), true);
  
  t = 2148400001.0000; 
  debugPrint(String.Format("2148400001 ERR %f", t), true);

  t = Maths.RaiseToPower(2.0 , 32.0) - 2146567296.0; 
  debugPrint(String.Format("2^32 - 2146567296 %f", t), true);
```


- 2148000000 2148000000.000000                             
- 2148400001 ERR 2148400128.000000                         
- 2^32 - 2146567296 2148400128.000000                      
- 4294967296.000000 - 2148400128.000000 = 2146567168.000000

**Comment by ericoporto** [comment]
I think the "integer" part of a float is limited to 23 bits of representation, so any attempt at more will just round it. 

2148400128 = 0b10000000000011011111110000000000

**Comment by jumpjack** [comment]
Possibily... but the I don't understand why this works properly:

float big = Maths.RaiseToPower(2.0 , 32.0)

Result:

4294967296.0000

**Comment by jumpjack** [comment]
> It's here: https://adventuregamestudio.github.io/ags-manual/ScriptKeywords.html#data-types
> 
> We can add a see-also call-out in the mentioned topics though.

Is there also a page listing the max number of "AGS elements"? Objects, sprites, walkbehinds,... A comparison of these numbers in different AGS versions would be also useful.

**Comment by ericoporto** [comment]
The right place to ask questions is the forums: https://www.adventuregamestudio.co.uk/forums/index.php

**Comment by jumpjack** [comment]
Sorry you're right, anyway I found it: https://adventuregamestudio.github.io/ags-manual/SystemLimits.html

**Comment by ivan-mogilko** [comment]
I think it's okay to ask for the general engine information here; but also the AGS manual has a [separate repository](https://github.com/adventuregamestudio/ags-manual), where one may open questions / requests if some information is missing or is difficult to find.

**Comment by ivan-mogilko** [comment]
I apologize for such a long delay. I think the information that was added to the manual since mostly addressed the confusion here:
https://adventuregamestudio.github.io/ags-manual/SystemLimits.html#integers-and-floats-limits

In short, unlike the integers that are strictly limited by certain min and max numbers, 32-bit floats do not have such a clear "ceiling", but instead they tend to loose precision while their integer part increases. Certain "even" numbers may still work, this is why Maths.RaiseToPower(2.0, 32.0) gives a correct result, other numbers are not guaranteed to be precise.

To be certain I repeated same tests in C program, and all the results were the same. So this is purely a limitation of 32-bit floats, not an engine bug.
